### DISCUSSION TITLE
Implement Refund Flow in Greenfield API

### DISCUSSION THREAD

**Discussion by mp21-org** [btcpayserver/btcpayserver#4181] [Technical Support] [answered]
As someone who wants to implement BTCPay as a payment module into WHMCS, I came across the topic of refunds.

We have this flow in the UI (rechable from the invoice detail view):

![grafik](https://user-images.githubusercontent.com/1374810/193281544-c85d5f92-f1d3-4cdd-9089-5f7129331d44.png)

However it is not accessable via the API.

For partial refunds, it would be nice to also have the choice between sending the sats amount at current rates or the partial sats amount at the rates when the original payment was made. (we will have to keep altcoins in mind here as well)

highly optional: The icing on top would be an option to send the pull payment link to the user via email. (reason: refunds in whmcs are issued by service personel or during cronjob runs: we can not show a link here that is meant for the enduser - might be able to do the email sending myself in the WHMCS module code)

I am pretty sure other integrations could benefit from a refund flow as well. Maybe @ndeet can share his 2 sats on the matter. ^^

**Comment by ndeet** [comment]
Yes most ecommerce systems have that functionality in some form. 

If I understand it correctly you think of creating a refund via the API, so you want to send some flags to the API so we can quickly create a pull payment like when creating the refund you specify if you want the full amount and if you want to have it nominated in fiat or crypto or a custom amount fiat/crypto? That will then return us the link to the pull payment? 

I wonder if this needs any separate refund endpoint or if we could do it with pull payments already.

**Comment by bolatovumar** [comment, marked as answer]
Hey @petzsch, I started a PR for this functionality here: https://github.com/btcpayserver/btcpayserver/pull/4238

**Comment by mp21-org** [comment]
This discussion can be closed when 1.7.1 is released. Thank you for the PR @bolatovumar 💚
